File: Go-K8s/mqsamputils/env.go
/**
 * © Copyright IBM Corporation 2020
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
**/

package mqsamputils

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"os"
	"strings"
)

var logger = log.New(os.Stdout, "Env: ", log.LstdFlags)

type Env struct {
	User             string `json:"APP_USER"`
	Password         string `json:"APP_PASSWORD"`
	QManager         string `json:"QMGR"`
	QueueName        string `json:"QUEUE_NAME"`
	ModelQueueName   string `json:"MODEL_QUEUE_NAME"`
	DynamicQueueName string `json:"DYNAMIC_QUEUE_PREFIX"`
	Host             string `json:"HOST"`
	Port             string `json:"PORT"`
	Channel          string `json:"CHANNEL"`
	Topic            string `json:"TOPIC_NAME"`
	KeyRepository    string `json:"KEY_REPOSITORY"`
	Cipher           string `json:"CIPHER"`
}

type MQEndpoints struct {
	Points []Env `json:"MQ_ENDPOINTS"`
}

var EnvSettings Env
var MQ_ENDPOINTS MQEndpoints

const FULL_STRING = -1

func init() {
	jsonFile, err := os.Open("env.json")
	defer jsonFile.Close()

	if err != nil {
		logger.Println(err)
		//return
	} else {
		logger.Println("Successfully Opened env.json")

		byteValue, _ := ioutil.ReadAll(jsonFile)
		json.Unmarshal(byteValue, &MQ_ENDPOINTS)
	}
	// The .json should have supplied the MQ Endpoints as an array.
	// If there are no elements, then EnvSettings will be default
	// initialised to be empty.
	if len(MQ_ENDPOINTS.Points) > 0 {
		EnvSettings = MQ_ENDPOINTS.Points[0]
	}

	environmentOverides()
}

func environmentOverides() {
	logger.Println("Looking for Environment Overrides")
	var s string

	overrides := map[string]*string{
		"APP_USER":             &EnvSettings.User,
		"APP_PASSWORD":         &EnvSettings.Password,
		"QMGR":                 &EnvSettings.QManager,
		"QUEUE_NAME":           &EnvSettings.QueueName,
		"MODEL_QUEUE_NAME":     &EnvSettings.ModelQueueName,
		"DYNAMIC_QUEUE_PREFIX": &EnvSettings.DynamicQueueName,
		"HOST":                 &EnvSettings.Host,
		"PORT":                 &EnvSettings.Port,
		"CHANNEL":              &EnvSettings.Channel,
		"TOPIC_NAME":           &EnvSettings.Topic,
		"KEY_REPOSITORY":       &EnvSettings.KeyRepository,
		"CIPHER":               &EnvSettings.Cipher,
	}

	for f, v := range overrides {
		logger.Printf("Overide for %s", f)
		s = os.Getenv(f)
		if s != "" {
			*v = s
		}
	}

	MQ_ENDPOINTS.Points = make([]Env, 1)
	MQ_ENDPOINTS.Points[0] = EnvSettings
}

func (Env) GetConnection(index int) string {
	if index == FULL_STRING {
		var connections []string
		for _, p := range MQ_ENDPOINTS.Points {
			connections = append(connections, p.Host+"("+p.Port+")")
		}
		return strings.Join(connections[:], ",")
	} else {
		p := MQ_ENDPOINTS.Points[index]
		return p.Host + "(" + p.Port + ")"
	}
}

func (Env) GetConnectionCount() int {
	return len(MQ_ENDPOINTS.Points)
}

func (Env) LogSettings() {
	logger.Println("Environment Settings are")
	logger.Printf("Username is (%s)\n", EnvSettings.User)
	//logger.Printf("Password is (%s)\n", EnvSettings.Password)
	logger.Printf("Queue Manager is (%s)\n", EnvSettings.QManager)
	logger.Printf("Queue Name is (%s)\n", EnvSettings.QueueName)
	logger.Printf("ModelQueue Name is (%s)\n", EnvSettings.ModelQueueName)
	logger.Printf("Host is (%s)\n", EnvSettings.Host)
	logger.Printf("Port is (%s)\n", EnvSettings.Port)
	logger.Printf("Connection is (%s)\n", EnvSettings.GetConnection(FULL_STRING))
	logger.Printf("Channel is (%s)\n", EnvSettings.Channel)
	logger.Printf("Topic is (%s)\n", EnvSettings.Topic)
	logger.Printf("Key Respository is (%s)\n", EnvSettings.KeyRepository)
	logger.Printf("Cipher (%s)\n", EnvSettings.Cipher)
}

File: amqp/dotnet/Env.cs
/****************************************************************************************/
/*                                                                                      */
/*                                                                                      */
/*  Copyright 2023 IBM Corp.                                                            */
/*                                                                                      */
/*  Licensed under the Apache License, Version 2.0 (the "License");                     */
/*  you may not use this file except in compliance with the License.                    */
/*  You may obtain a copy of the License at                                             */
/*                                                                                      */
/*  http://www.apache.org/licenses/LICENSE-2.0                                          */
/*                                                                                      */
/*  Unless required by applicable law or agreed to in writing, software                 */
/*  distributed under the License is distributed on an "AS IS" BASIS,                   */
/*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.            */
/*  See the License for the specific language governing permissions and                 */
/*  limitations under the License.                                                      */
/*                                                                                      */
/*                                                                                      */
/****************************************************************************************/


namespace ibmmq_amqp_samples
{
    class Env
    {
        public class MQEndPoints
        {
            public List<ConnVariables>? mq_endpoints;
        }

        public class ConnVariables
        {
            public string host = null;
            public string qmgr = null;
            public int port = 0;
            public string channel = null;
            public string queue_name = null;
            public string model_queue_name = null;
            public string topic_name = null;
            public string app_user = null;
            public string app_password = null;
            public string cipher_suite = null;
            public string key_repository = null;
            public bool is_managed = true;
            public void dump()
            {
                Console.WriteLine("hostname {0} ", host);
                Console.WriteLine("port {0} ", port);
                Console.WriteLine("qmgr {0} ", qmgr);
                Console.WriteLine("channel {0} ", channel);
                Console.WriteLine("queue {0} ", queue_name);
                Console.WriteLine("topic {0} ", topic_name);
                Console.WriteLine("app_user {0} ", app_user);
                Console.WriteLine($"is_managed {is_managed}");
                // Console.WriteLine("app_password {0} ", app_password);
                Console.WriteLine("cipherSpec {0} ", cipher_suite);
                Console.WriteLine("sslKeyRepository{0} ", key_repository);
            }
        }

        private MQEndPoints points = null;
        private ConnVariables conn = null;

        internal ConnVariables Conn { get => conn; set => conn = value; }

        private bool AtLeast(int i)
        {
            if (i >= 0 && points != null && points.mq_endpoints != null && points.mq_endpoints.Count >= i)
            {
                return true;
            }
            return false;
        }

        public int NumberOfConnections()
        {
            if (AtLeast(1))
            {
                return points.mq_endpoints.Count;
            }
            return 0;
        }

        private ConnVariables EndPoint(int i)
        {
            if (AtLeast(i))
            {
                return points.mq_endpoints[i];
            }
            return null;
        }

        public IEnumerable<ConnVariables> GetEndpoints()
        {
            int count = NumberOfConnections();
            for (int i = 0; i < count; i++)
            {
                yield return points.mq_endpoints[i];
            }
        }

        public string BuildConnectionString()
        {
            List<string> connList = new List<string>();

            if (AtLeast(1))
            {
                foreach (var c in points.mq_endpoints)
                {
                    if (c.host != null && c.port != 0)
                    {
                        connList.Add(c.host + "(" + c.port.ToString() + ")");
                    }
                }
                return string.Join(",", connList);
            }

            return "";
        }

        public bool ReadJson(string getEnv)
        {
            bool isSet = false;

            using (StreamReader r = new StreamReader(getEnv))
            {
                Console.WriteLine("File found");
                string json = r.ReadToEnd();

                points = JsonConvert.DeserializeObject<Env.MQEndPoints>(json);

                if (points != null && points.mq_endpoints != null && points.mq_endpoints.Count > 0)
                {
                    Conn = points.mq_endpoints[0];
                    Conn.dump();
                    isSet = true;
                }
            }

            return isSet;
        }

        public bool EnvironmentIsSet()
        {
            bool isSet = false;
            try
            {
                Console.WriteLine("Looking for file");

                // Getting the path from env_variable : ENV_FILE
                string getEnv = Environment.GetEnvironmentVariable("ENV_FILE");

                if (getEnv == null || !ReadJson(getEnv))
                {
                    // Use the Default env.json in the project folder
                    getEnv = @"../../../env.json";
                    if (ReadJson(getEnv))
                    {
                        Console.WriteLine("Using Default env.json");
                        isSet = true;
                    }
                    else
                    {
                        Console.WriteLine("MQ Settings not found, unable to determine connection variables");
                    }
                }
                else
                {
                    Console.WriteLine("Using externally Provided env.json");
                    isSet = true;
                }

                Console.WriteLine("");
                return isSet;
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception caught: {0}", e);
                Console.WriteLine(e.GetBaseException());
                return isSet;
            }
        }
    }
}
